### PR TITLE
✨ Reencode task for worker

### PR DESCRIPTION
**Pull request by pajowu**
Reencode everything to 128k cbr mp3 and m4a

### CODE REVIEW

**Comment by anuejn** [comment]
Do we really want to do this in the backend?.
I have the vague recollection, that we discussed to do it in the frontend, because it saves us compute and the users bandwidth.

**Comment by pajowu** [comment]
Yes and no. I think it makes sense to do this in the frontend too, but I think we should have to option in the backend as well. a) to make sure the file is in the right format and b) to handle cases where the conversion in the frontend is not possible (e.g. files larger than 2GB, browsers without wasm)

File: worker/pyproject.toml [hunk 1]
```diff
@@ -24,6 +24,7 @@ dependencies = [
     "torchvision==0.14.1",
     "hmmlearn @ git+https://github.com/hmmlearn/hmmlearn@e22127765b0ca70d01976df2a369667756b8445e",             # hmmlearn version which actually compiled 🙃. Needs to be specified here instead of in pdm.resolution.overrides because pdm crashes otherwise 🙃🙃
     "lightning-lite>=1.8.6",
+    "ffmpeg-python>=0.2.0",
```
Comment (anuejn) [inline worker/pyproject.toml:27]: could you add ffmpeg to our shell.nix or does this bring its own ffmpeg?

**Review by anuejn — COMMENTED** [review]
I am fine with merging transcription at the worker since it is an improvement from the status quo :). However I would like to investigate transcoding at the frontend in the future.

**Comment by anuejn** [comment]
fair points :)

**Comment by anuejn** [comment]
Ah we probably also need a way to inform the frontend, that the audio file changed. -> Future

**Review by anuejn — APPROVED** [review]

File: backend/transcribee_backend/routers/document.py [hunk 2]
```diff
@@ -57,6 +69,7 @@ def create_default_tasks_for_document(session: Session, document: Document):
         task_type=TaskType.TRANSCRIBE,
         task_parameters={"lang": "auto", "model": "base"},
         document_id=document.id,
+        dependencies=[reencode_task],
```
Comment (anuejn) [inline backend/transcribee_backend/routers/document.py:72]: is this needed?

File: backend/transcribee_backend/routers/document.py [hunk 3]
```diff
@@ -191,3 +204,47 @@ async def websocket_endpoint(
...
+    tags: list[str] = Form(),
+    file: UploadFile = File(),
+    session: Session = Depends(get_session),
+    authorized_worker: Worker = Depends(get_authorized_worker),
+) -> ApiDocument:
+    statement = select(Task).where(
+        Task.document_id == document_id, Task.task_type == TaskType.REENCODE
+    )
+    task = session.exec(statement).one_or_none()
+    if task is None:
```
Comment (anuejn) [inline backend/transcribee_backend/routers/document.py:221]: for what do we need the task here?

File: frontend/src/editor/player.tsx [hunk 4]
```diff
@@ -18,7 +18,16 @@ export function PlayerBar({
   documentContent: Descendant[];
 }) {
   const { data } = useGetDocument({ document_id: documentId });
-  const audioFile = data?.media_files[0]?.url;
+  let audioFile = data?.media_files[0]?.url;
+  const audioElement = document.createElement('audio');
+  data?.media_files.forEach((media_file) => {
+    if (
+      media_file.tags.indexOf('original') == -1 &&
+      audioElement.canPlayType(media_file.content_type) == 'probably'
```
Comment (anuejn) [inline frontend/src/editor/player.tsx:26]: nice :D

File: backend/transcribee_backend/routers/document.py (hunk 2 above)
Comment (pajowu) [reply backend/transcribee_backend/routers/document.py:72]: Yes, `load_audio` only supports a small number of formats (eg doesn’t do mp4) -> ffmpeg converts them to a safe format

File: backend/transcribee_backend/routers/document.py (hunk 3 above)
Comment (pajowu) [reply backend/transcribee_backend/routers/document.py:221]: This is a permission check: the worker is only allowed to upload media if it is assigned to the reencode task

File: backend/transcribee_backend/routers/document.py (hunk 2 above)
Comment (anuejn) [reply backend/transcribee_backend/routers/document.py:72]: Hm... can we skip the dependency if we are likely to be able to load it directly? Or use a library that can load more formats? This seems like a downgrade performance-wise :/

Comment (pajowu) [reply backend/transcribee_backend/routers/document.py:72]: I'd prefer not to, as it also calms my mind a bit to know that (after it went through ffmpeg, which is it list fuzzed a bit), the files are in a sortof-safe state. Also the reencoding is incredibly fast

File: backend/transcribee_backend/routers/document.py [hunk 5]
```diff
@@ -191,3 +204,47 @@ async def websocket_endpoint(
...
+        created_at=now_tz_aware(),
+        changed_at=now_tz_aware(),
+        document_id=task.document_id,
+        file=stored_file,
+        content_type=magic.from_descriptor(file.file.fileno(), mime=True),
+    )
+
+    session.add(media_file)
+    session.commit()
+    session.refresh(media_file)
```
Comment (rroohhh) [inline backend/transcribee_backend/routers/document.py:240]: This refresh is also unnecessary. Doesn't `sqlmodel` automatically expire on commit and then fetch the most recent data when its used later?

**Review by rroohhh — CHANGES_REQUESTED** [review]

File: backend/transcribee_backend/routers/document.py (hunk 5 above)
Comment (pajowu) [reply backend/transcribee_backend/routers/document.py]: Nice, thanks for all the three comments. I think I just commited the `add`/`commit`/`refresh`-code somewhere from the sqlmodel docs without thinking about it again

**Review by rroohhh — APPROVED** [review]
`load_audio` should probably be moved to the `ffmpeg-python` package at some point aswell

**Review by rroohhh — APPROVED** [review]
`load_audio` should probably be moved to the `ffmpeg-python` package at some point aswell

**Pull request merged**